src/couch_replicator.erl

```diff
@@ -103,935 +68,272 @@ replicate(PostBody, Ctx) ->
+            couch_log:warning("replicator scheduler: ~p was already added", [Rep])
+        end,
+        ok;
+    #rep{doc_id = DocId} ->
+        ok;
+    #rep{db_name = DbName, doc_id = OtherDocId} ->
+        couch_log:notice("The replication specified by the document `~s` already started"
+            " triggered by the document `~s`", [DocId, OtherDocId]),
+        maybe_tag_rep_doc(DbName, DocId, RepDoc, ?l2b(BaseId))
+    end,
+    ok.
 
 
-get_pending_count_int(#rep_state{source = #httpdb{} = Db0}=St) ->
-    {_, Seq} = St#rep_state.highest_seq_done,
-    Db = Db0#httpdb{retries = 3},
-    case (catch couch_replicator_api_wrap:get_pending_count(Db, Seq)) of
-    {ok, Pending} ->
-        Pending;
+-spec maybe_tag_rep_doc(binary(), binary(), {[_]}, binary()) -> ok.
+maybe_tag_rep_doc(DbName, DocId, {RepProps}, RepId) ->
+    case get_json_value(<<"_replication_id">>, RepProps) of
+    RepId ->
+        ok;
     _ ->
-        null
-    end;
-get_pending_count_int(#rep_state{source = Db}=St) ->
-    {_, Seq} = St#rep_state.highest_seq_done,
-    {ok, Pending} = couch_replicator_api_wrap:get_pending_count(Db, Seq),
-    Pending.
-
-
-update_task(State) ->
-    #rep_state{
-        current_through_seq = {_, ThroughSeq},
-        highest_seq_done = {_, HighestSeq}
-    } = State,
-    couch_task_status:update(
-        rep_stats(State) ++ [
-        {source_seq, HighestSeq},
-        {through_seq, ThroughSeq}
-    ]).
+        couch_replicator_docs:update_doc_replication_id(DbName, DocId, RepId)
+    end.
 
 
-rep_stats(State) ->
-    #rep_state{
-        committed_seq = {_, CommittedSeq},
-        stats = Stats
-    } = State,
+-spec remove_jobs(binary(), binary()) -> ok.
+remove_jobs(DbName, DocId) ->
+    LogMsg = "Stopped replication `~s` , replication document `~s`",
     [
-        {revisions_checked, couch_replicator_stats:missing_checked(Stats)},
-        {missing_revisions_found, couch_replicator_stats:missing_found(Stats)},
-        {docs_read, couch_replicator_stats:docs_read(Stats)},
-        {docs_written, couch_replicator_stats:docs_written(Stats)},
-        {changes_pending, get_pending_count(State)},
-        {doc_write_failures, couch_replicator_stats:doc_write_failures(Stats)},
-        {checkpointed_source_seq, CommittedSeq}
-    ].
+        begin
+            couch_replicator_scheduler:remove_job(RepId),
+            couch_log:notice(LogMsg, [pp_rep_id(RepId), DocId])
+        end || RepId <- find_jobs_by_doc(DbName, DocId)
+    ],
+    ok.
+
+
+% pretty-print replication id
+-spec pp_rep_id(#rep{} | rep_id()) -> string().
+pp_rep_id(#rep{id = RepId}) ->
+    pp_rep_id(RepId);
+pp_rep_id({Base, Extension}) ->
+    Base ++ Extension.
+
+
+
+% TODO: make this a function in couch_replicator_scheduler API
+-spec clean_up_replications(binary()) -> ok.
+clean_up_replications(DbName) ->
+    RepIds = find_jobs_by_dbname(DbName),
+    [couch_replicator_scheduler:remove_job(RepId) || RepId <- RepIds],
+    ok.
```

Comment:
Another nit: You can replace the above two lines with `lists:foreach(fun couch_replicator_scheduler:remove_job/1, RepIds).`